### ISSUE TITLE
OH3: Switch item with linked channel and hysteresis profile leads to error on the item

### ISSUE THREAD

**Issue by FSeidinger** [openhab/openhab-core#2037]
I'm using openHAB 3.0 with docker on a linux host system.

My use case is an alarm indicating that the humidity in my bathroom is over a certain upper value. When this happens the bathroom fan is turned on to bring the humidity down. The fan should stay on until the humidity drops below a certain lower value.

For this I configured a switch item carrying the alarm value and linked it to the humidity sensor with a channel configured for the hysteresis profile.

The expected outcome should be a switch item handling the alarm value with the hysteresis in place. The current outcome is that the item goes into an error state.

The item configuration is:

![image](https://user-images.githubusercontent.com/4890418/103382284-c3a56380-4aee-11eb-85e4-2275a2a95f19.png)

The channel link configuration is:

![image](https://user-images.githubusercontent.com/4890418/103382325-ef284e00-4aee-11eb-9825-12a0d80d3404.png)

As you can see in the screen shots, the state is "Err". The probable cause might be the following log entry:

```
2020-12-30 00:41:22.012 [WARN ] [e.internal.SseItemStatesEventBuilder] - Exception while formatting value 'OFF' of item Badezimmer_Sicherheit_Feuchtigkeitsalarm with format '%.0f': f != java.lang.String
```

**Comment by cweitkamp** [comment]
The problem is that the Switch Item inherits its pattern to format the value by the linked Channel which provides a Number Item. The easy way to omit the occuring warning is to define a Metadata > State Description > Pattern (e.g. `%s`) for the Switch Item.

I am not sure if it will be the better option to add a fix for this in the `SseItemStatesEventBuilder`.

//CC @ghys wdyt?

**Comment by ghys** [comment]
Yes I think it could be best to try/catch transformation & pattern formatting errors and just provide the raw state if there's an exception. It's better than this "Err" label.

**Comment by FSeidinger** [comment]
I investigated her a little and think that the problem is with the current handling/implementation of the profiles. The profiles are only known on the thing level through the `org.openhab.core.thing.internal.CommunicationManager`. If a profile is applied a state/command conversion is delegated to the profile and the translated state/command is then send to the item.

The item on the other hand tries to format the state by taking the `org.openhab.core.thing.internal.ChannelStateDescriptionProvider`. The description provider doesn't know about the profile and takes the state description of the linked channel, which is a number and therefore results in the format mismatch between the item type (switch) and the channel type (number).

The best way would be that the profile is assigned to the link channel and therefore the state description can take the transformed state into account when formatting the value.

**Comment by FSeidinger** [comment]
> I am not sure if it will be the better option to add a fix for this in the `SseItemStatesEventBuilder`.
> //CC @ghys wdyt?

The `SseItemStatesEventBuilder` is not the problem here, I guess. The event builder simply take the state description and the format from there to build a suitable event.

As I mentioned above, the state description provider should to the job to detect an assigned profile and alter the state description accordingly.

**Comment by FSeidinger** [comment]
> The problem is that the Switch Item inherits its pattern to format the value by the linked Channel which provides a Number Item. The easy way to omit the occuring warning is to define a Metadata > State Description > Pattern (e.g. `%s`) for the Switch Item.

I tested this workaround with the `%s` pattern and the switch value is then correctly formatted as either `ON` or `OFF`.

**Comment by Rossko57** [comment]
> As I mentioned above, the state description provider should do the job to detect an assigned profile and alter the state description accordingly.

If considering profiles in general, some fairly arbitrary transformations can be achieved by the user.  It's not possible to guess what those might be; difficult to solve.
An 'educated guess' of some desired state description could be made by looking at target Item type.
It might work better to just suppress channel-suggested state description - but only  where channel type does not match Item type.

**Comment by FSeidinger** [comment]
> If considering profiles in general, some fairly arbitrary transformations can be achieved by the user. It's not possible to guess what those might be; difficult to solve.
> An 'educated guess' of some desired state description could be made by looking at target Item type.
> It might work better to just suppress channel-suggested state description - but only where channel type does not match Item type.

Ah, understood.

In this case the profile type itself should carry an altered state description. For the profile types standard, scale, follow and hysteresis, the state description can be set from the profile itself. For the profile types map and js, the user should supply a suitable state description.

If the profile would be linked to the channel, the state description can then be taken from the profile instance for formatting.

**Comment by Rossko57** [comment]
This community thread is I think another manifestation of the same problem
https://community.openhab.org/t/oh-3-scale-profile-exception-while-formatting-value/111022

in this case, transforming a Quantity Type channel with default presentation into a string Item.

**Comment by Rossko57** [comment]
A user Brian Warner has gone to the trouble of investigating deeper on forum post 121891 linked above.
Conclusion, my idea of the problem mechanism at the profile action was completely wrong and @cweitkamp correctly focused on Item pattern.

When a channel link is made to an Item, **if** the binding provides a default 'pattern' it is **forced** onto the Item.
No account is taken of any existing pattern, or Item type differences.
In this case,  a battery charge channel forces format %.0f  onto a timestamp DateTime Item and of course causes errors.

A minimal fix would simply to be a good neighbour - if an Item already has any existing pattern DO NOT force-feed a channel suggested default pattern. 
I think this should be the case in general, never mind profiles in particular.

A more intelligent addition would to look at "target Item" and "source channel" types as well.  If there is gross mis-match, again do not apply channel suggested default to Item even if it is currently blank.

These are core behaviour changes, and not really profile specific - profile use just shows up the flaw in the defaulting process.
This might seem a trivial issue but it comes up regularly, each time in a slightly different disguise.

**Comment by cweitkamp** [comment]
I agree with https://github.com/openhab/openhab-core/issues/2037#issuecomment-753381920 and an easy way to fix this is to change the following line 157

https://github.com/openhab/openhab-core/blob/a82d3effb7bf268249ec066bb5c67613dc08a312/bundles/org.openhab.core.io.rest.sse/src/main/java/org/openhab/core/io/rest/sse/internal/SseItemStatesEventBuilder.java#L153-L158

to

```java
         displayState = state.toString();
```

and suppress the warning.

**Comment by Rossko57** [comment]
Uhh, won't that stop 'pattern' doing what it is supposed to do, format a state for display?

**Comment by brianwarner** [comment]
@Rossko57 I think it should work out as expected, as it would only fall back in the failure case? One sec and I'll open a PR and we can stare at it.

**Comment by brianwarner** [comment]
Sorry for all the noise on the branch, I had to force push a few times to correct some mistakes in the commit message. The last one is the only one you need to look at.

@cweitkamp I changed the warn to info, thinking it will still provide meaningful feedback to someone who goofs up their pattern metadata, but is easier to ignore when someone just wants to use "Timestamp on" in its default state.

**Comment by Rossko57** [comment]
I'm not at all sure suppressing the error is the correct approach.
The root issue is the framework auto-feeding the incorrect pattern in the first place.

Note that 'pattern' also gets used for other non-display purposes, e.g defining default UoM units.

**Comment by Rossko57** [comment]
Ah, here's another consequence that we have overlooked so far.
The "forced suggestion" of a 'pattern' from the binding may _also_ be accompanied by `state options `(e.g.open/close).
Manually fixing the 'pattern' alone is not enough then.

**Comment by MyRaceData** [comment]
This issue seems to be about hysteresis profile but the issue seems to effect timestamp profiles as well not being able to be formatted as mentioned in the above community post. [linked here for reference](https://community.openhab.org/t/datetime-format-with-oh3-items-not-working/130130) The items linked to are zwave contact sensors and three different users (including myself) with several different brands of contact sensors (two different brands myself) all report same issue
just to document
DateTime type item with format pattern in metadata
![Screenshot from 2021-12-18 12-28-32](https://user-images.githubusercontent.com/25418996/146680459-f6cf4bfa-b75f-4629-b406-c9b8a1225d86.png)
output in widget on page
![Screenshot from 2021-12-18 12-27-03](https://user-images.githubusercontent.com/25418996/146680480-56087387-4088-40c2-9adf-6ebb8f17a7bd.png)
unlink timestamp profile
![Screenshot from 2021-12-18 12-29-24](https://user-images.githubusercontent.com/25418996/146680501-e268ac14-ec4b-4ba3-84d5-3ab671ec82a0.png)
format immediately applied to widget
![Screenshot from 2021-12-18 12-29-51](https://user-images.githubusercontent.com/25418996/146680519-5ce55fc4-f948-4fc8-a48e-fe36593e93c9.png)

**Comment by brianwarner** [comment]
Confirming I think I'm seeing the same. In my case I have my Z-Wave devices' channel(s) linked to a single "last update" item for each device. Where it's a single channel the state metadata is used correctly and I can format the timestamp. When there are multiple channels it falls back to the default unformatted timestamp. I looked into it briefly a few months back so I'm sorry if the description is a bit hazy, but my guess is that when there are multiple links on an item that doesn't use the default profile, the payload is delivered in a way that can't be formatted. Prior to the change that @Rossko57 and I worked on it probably would have reported "Err" in the UI. Now it falls back to an unformatted datetime so that if something goes wrong with the aesthetics it's still providing useful info by default. At the time we thought there may be a deeper underlying cause but neither of us could find it.

**Comment by Rossko57** [comment]
It's going to require some kind of framework policy decision.  Should linking an Item to a channel always allow the binding to 'force-feed' presentation metadata pattern and options?
If not, why not?  When there are multiple links, perhaps?  And/or when target Item type does not match channel type?

Or add an option to the linking process, supply by default but allow denial by user selection?
